ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +158,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+				headServices.Items[0].Name,
+				headServices.Items[0].Namespace,
+				dashboardPort)
+			return dashboardURL, nil
+		}
+
+		// This should never happen.
+		// We add the protection here just in case controller has race issue or user manually create service with same label.
+		if len(headServices.Items) > 1 {
+			r.Log.Info("reconcileServices ", "Duplicates head service found", len(headServices.Items))
+			return "", fmtErrors.Errorf("Duplicates head service found %v", len(headServices.Items))
+		}
+	}
+
+	// Create head service if there's no existing one in the cluster.
+	if headServices.Items == nil || len(headServices.Items) == 0 {
+		return "", fmtErrors.Errorf("No head service found")
+	}
+
+	return "", nil
+}
+
+func (r *RayServiceReconciler) checkIfNeedSubmitServeDeployment(rayServiceInstance *rayv1alpha1.RayService, request ctrl.Request) bool {
+	existConfig, exist := r.ServeDeploymentConfigMap[request.NamespacedName]
+
+	shouldUpdate := false
+
+	if !exist || !reflect.DeepEqual(existConfig, rayServiceInstance.Spec) || len(rayServiceInstance.Status.ServeStatuses.Statuses) != len(existConfig.ServeConfigSpecs) {
+		shouldUpdate = true
+	}
+	r.Log.Info("shouldUpdate value", "shouldUpdate", shouldUpdate)
+
+	log.V(1).Info("status check", "len(rayServiceInstance.Status.ServeStatuses.Statuses) ", len(rayServiceInstance.Status.ServeStatuses.Statuses), "len(existConfig.ServeConfigSpecs)", len(existConfig.ServeConfigSpecs))
+
+	return shouldUpdate
+}
+
+func (r *RayServiceReconciler) updateServeDeployment(rayServiceInstance *rayv1alpha1.RayService, rayDashboardClient utils.RayDashboardClient, request ctrl.Request) error {
+	r.Log.Info("shouldUpdate")
+	if err := rayDashboardClient.UpdateDeployments(rayServiceInstance.Spec.ServeConfigSpecs); err != nil {
+		r.Log.Error(err, "fail to update deployment")
+		return err
+	}
+
+	r.ServeDeploymentConfigMap[request.NamespacedName] = rayServiceInstance.Spec
+	return nil
+}
+
+func (r *RayServiceReconciler) getAndCheckServeStatus(rayServiceInstance *rayv1alpha1.RayService, dashboardClient utils.RayDashboardClient) (bool, error) {
+	var serveStatuses *rayv1alpha1.ServeStatuses
+	var err error
+	if serveStatuses, err = dashboardClient.GetDeploymentsStatus(); err != nil {
+		r.Log.Error(err, "fail to get deployment status")
+		return false, err
+	}
+
+	statusMap := make(map[string]rayv1alpha1.ServeStatus)
+
+	for _, status := range rayServiceInstance.Status.ServeStatuses.Statuses {
+		statusMap[status.Name] = status
+	}
+
+	isHealthy := true
+	for i := 0; i < len(serveStatuses.Statuses); i++ {
+		serveStatuses.Statuses[i].HealthLastUpdateTime = metav1.Now()
+		if serveStatuses.Statuses[i].Status != "HEALTHY" {
+			prevStatus, exist := statusMap[serveStatuses.Statuses[i].Name]
+			if exist {
+				if prevStatus.Status != "HEALTHY" {
+					serveStatuses.Statuses[i].HealthLastUpdateTime = prevStatus.HealthLastUpdateTime
+
+					if time.Since(prevStatus.HealthLastUpdateTime.Time).Seconds() > 60 {
```

Comment:
Is the 60 second timeout something we'd want to be configurable?
If not, we should probably define it as a constant somewhere.